noxfile_utils.py

```diff
@@ -0,0 +1,305 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Utility functions and decorators for managing and customizing Nox sessions.
+
+It includes functionality for handling session-specific metadata, like determining
+whether a session should be skipped based on the changes in the repository.
+
+Environment Variables:
+
+- `[{PROJECT_PREFIX}_]CI_NOX_RUN_ONLY_IF_CHANGED_FROM`: Specifies the
+  commit to compare changes against.
+- `[{PROJECT_PREFIX}_]CI_NOX_VERBOSE`: Enables verbose mode for debugging
+  session behavior.
+
+"""
+
+from __future__ import annotations
+
+import fnmatch
+import functools
+import itertools
+import os
+import pathlib
+import subprocess
+import sys
+import types
+import unittest
+from collections.abc import Callable, Iterable, Sequence
+from typing import Any, Final, TypeAlias
+import unittest.mock
+
+import nox
+import nox.registry
+
+ENV_VAR_PREFIX: Final = f"{prefix}_" if (prefix := os.environ.get("CI_NOX_PREFIX", "")) else ""
+ENV_VAR_RUN_ONLY_IF_CHANGED_FROM: Final = f"{ENV_VAR_PREFIX}CI_NOX_RUN_ONLY_IF_CHANGED_FROM"
+ENV_VAR_VERBOSE: Final = f"{ENV_VAR_PREFIX}CI_NOX_VERBOSE"
+
+COMMIT_SPEC_FOR_CHANGES: Final = os.environ.get(ENV_VAR_RUN_ONLY_IF_CHANGED_FROM, "")
+VERBOSE: Final = os.environ.get(ENV_VAR_VERBOSE, "").lower() in [
+    "1",
+    "on",
+    "true",
+]
+
+_changed_files_from_commit: dict[str, list[str]] = {}
```

Comment:
Another question: what is the live time of these caches? One pipeline? one call to nox?

Reply:
It is one call to nox, but since `_changed_files_from_commit` is only populated in the session run, then is effectively per session, which doesn't make any sense. Good catch, I 'll remove it.